### PR TITLE
Disable request body data rate limits per HTTP/2 stream

### PR DESCRIPTION
**Pull request by halter73**
Addresses #7635

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs [hunk 1]
```diff
@@ -227,6 +228,12 @@ bool IHttpBodyControlFeature.AllowSynchronousIO
             }
         }
 
+        MinDataRate IHttpMinRequestBodyDataRateFeature.MinDataRate
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs]: Can this be in the Http1Connection class still?

**Review by jkotalik — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs [hunk 2]
```diff
@@ -30,5 +34,19 @@ IHeaderDictionary IHttpResponseTrailersFeature.Trailers
         }
 
         int IHttp2StreamIdFeature.StreamId => _context.StreamId;
+
+        MinDataRate IHttpMinRequestBodyDataRateFeature.MinDataRate
+        {
+            get => throw new NotSupportedException(CoreStrings.Http2MinDataRateNotSupported);
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs]: The getter should return null if that's the only allowed value.

**Review by Tratcher — APPROVED** [review]

**Comment by JamesNK** [comment]
I'd like to understand how we'll use this in gRPC.

I believe we only want to disable the rate limit on gRPC requests. We could be more specific and only disable the rate limit on client or duplex streaming request. Would that be done by setting this feature on the HttpContext with a null MinDataRate on only gRPC client/duplex streaming requests?

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs (hunk 1 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs]: Yep.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs (hunk 2 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.FeatureCollection.cs]: The thing is we generally are enforcing a rate limit for HTTP/2, it's just per connection, not per stream. So I don't think it makes sense to return null, but I don't think it makes sense to return the connection limit either.

**Comment by halter73** [comment]
> Would that be done by setting this feature on the HttpContext with a null MinDataRate on only gRPC client/duplex streaming requests?

Exactly. Something like https://github.com/aspnet/AspNetCore/pull/10355/files#diff-2d9b5c655816cc6e1ea41260d95cd33eR815

**Comment by JamesNK** [comment]
Ah, the `IHttpMinRequestBodyDataRateFeature` will always be set on a HTTP/2 request, and it is a matter of getting it and setting it to null in the appropriate requests.

**Comment by halter73** [comment]
> What happens if you have a HTTP/2 request and someone defines there own IHttpMinRequestBodyDataRateFeature implementation and has a non-null MinDataRate?

That's the prerogative of whoever defines their own IHttpMinRequestBodyDataRateFeature. I doubt a custom IHttpMinRequestBodyDataRateFeature would be able to do much with a non-null MinDataRate and an HTTP/2 request unless is measured the rate of the request body itself.

**Comment by halter73** [comment]
I expect most people who replace a IHttpMinRequestBodyDataRateFeature would be replacing it with a facade that would call the "real" MinDataRate setter which would throw a NotSupportedException given a non-null rate with HTTP/2.

**Comment by JamesNK** [comment]
Ok. I'm not super familiar with this space, just want to make sure it is thought about.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**